code/datums/components/egg_layer.dm

```diff
@@ -0,0 +1,84 @@
+/**
+ * # egg layer component!
+ *
+ * Component that manages how many eggs to lay, what can be fed to the mob to make them lay more, and what is actually laid.
+ * Since the only real interaction with the component is an attackby, the nice part is that we're able to make this an atom level proc.
+ * egg_layer will loudly fail if you do not provide it the arguments, as to encourage explicicy(?)
+ */
+/datum/component/egg_layer
+	/// item laid by the mob
+	var/egg_type
+	/// items that can be fed to the mob to make it lay more eggs
+	var/list/food_types
+	/// messages sent when fed
+	var/list/feedMessages
+	/// messages sent when laying an egg
+	var/list/layMessages
+	/// kind of a weird boolean, but basically egg layers usually need to make their eggs process to let them grow. and some don't want that to happen too!
+	var/process_eggs
```

Comment:
If this is what I think this is, then the consumers should just start processing in the callback.